software/contrib/bouncing_pixels.py

```diff
@@ -0,0 +1,667 @@
+            .with_unlocked_knob("impulse_speed")
+            .with_locked_knob("width", initial_percentage_value=saved_state.get("width", 1.0))
+            .build()
+        )
+
+        # Initialise some internal state variables
+        self.b1_pressed = None
+        self.b2_pressed = None
+        self.ball_count = 1
+        self.time_factor = 1
+
+        # Trackers of inputs
+        # -1 as a starting value means they always get updated on first poll.
+        self.speed_input = -1.0
+        self.width_input = -1.0
+        self.ball_count_input = -1.0
+        self.impulse_speed_input = -1.0
+        self.ain_input = -1.0
+
+        # We store analogue input values in these. Although we only change change one as determined by
+        # config.ain_function, we do need to reference them all when their respective knobs change.
+        self.speed_ain_term = 0.0
+        self.width_ain_term = 0.0
+        self.ball_count_ain_term = 0.0
+        self.impulse_speed_ain_term = 0.0
+
+        # Initialise input events
+        self.on_speed_input = Event(self.mark_state_unsaved, self.apply_speed)
+        self.on_width_input = Event(self.mark_state_unsaved, self.apply_width)
+        self.on_ball_count_input = Event(self.mark_state_unsaved, self.apply_ball_count)
+        self.on_impulse_speed_input = Event(self.mark_state_unsaved, self.apply_impulse_speed)
+        self.on_ain_input = Event(
+            # To avoid excess compute spent on string formatting or dict lookups in the main loop, we keep an assortment of
+            # set and apply functions so that we can assign them once and for all here.
+            getattr(self, f"update_{self.config.ain_function}_ain_term"),
+            getattr(self, f"apply_{self.config.ain_function}"),
+        )
+
+        # Create the playing field and gate abstractions
+        self.arena = Arena(self.config)
+        self.balls = [
+            Ball(arena=self.arena, config=self.config) for _ in range(self.config.ball_count_max)
+        ]
+        hold_lengths = [
+            self.config.gate_hold_length_top,
+            self.config.gate_hold_length_left,
+            self.config.gate_hold_length_right,
+            self.config.gate_hold_length_bottom,
+            self.config.gate_hold_length_any,
+            self.config.gate_hold_length_corner,
+        ]
+        self.gates = [Gate(cv, hold_length) for cv, hold_length in zip(cvs, hold_lengths)]
+
+        # Register external event handlers.
+        b1.handler(self.b1_rising)
+        b1.handler_falling(self.b1_falling)
+        b2.handler(self.b2_rising)
+        b2.handler_falling(self.b2_falling)
+        din.handler(getattr(self, self.config.din_function))
+
+        for ball in self.balls:
+            ball.on_collide += self.report_collision
+
+    @classmethod
+    def config_points(cls):
+        return [
+            configuration.floatingPoint(
+                "poll_frequency", minimum=5.0, maximum=120.0, default=20.0, danger=True
```

Comment:
EuroPi's unwritten-but-should-be-written styleguide for config points usually treats them as constants, and assigns them `ALL_CAPS` names accordingly; see `europi_config.MAX_OUTPUT_VOLTAGE`. For a contrib script it's not required, but if you _really_ want to stay consistent you might consider renaming your config points.

Reply:
Noted and addressed, thank you :)